src/compiler/checker.ts

```diff
@@ -16535,6 +16517,89 @@ namespace ts {
             }
         }
 
+        // No signature was applicable. We have already reported the errors for the invalid signature.
+        // If this is a type resolution session, e.g. Language Service, try to get better information than anySignature.
+        function getCandidateForOverloadFailure(
+            node: CallLikeExpression,
+            candidates: Signature[],
+            args: ReadonlyArray<Expression>,
+            hasCandidatesOutArray: boolean,
+        ): Signature {
+            Debug.assert(candidates.length > 0); // Else should not have called this.
+
+            // Normally we will combine overloads. Skip this if they have type parameters since that's hard to combine.
+            // Don't do this if there is a `candidatesOutArray`,
+            // because then we want the chosen best candidate to be one of the overloads, not a combination.
+            if (!hasCandidatesOutArray && candidates.length > 1 && !candidates.some(c => !!c.typeParameters)) {

```

Comment:
1. I think the pick-longest-signature code would benefit from being in a separate function as well.
2. I think this condition would read better inverted in that case.
3. If you don't do (1) and (2), I would recommend pushing the negation inside `some` and inverting it to `candidate.every(c => !c.typeParameters)`. ("every not" is easier to read as "none" than "not some")